### PR TITLE
[SPARK-35703][SQL][FOLLOWUP] Only eliminate shuffles if partition keys contain all the join keys

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/32875 . Basically https://github.com/apache/spark/pull/32875 did two improvements:
1. allow bucket join even if the bucket hash function is different from Spark's shuffle hash function
2. allow bucket join even if the hash partition keys are subset of join keys.

The first improvement is the major target for implementing the SPIP "storage partition join". The second improvement is kind of a consequence of the framework refactor, which is not planned.

This PR is to disable the second improvement by default, which may introduce perf regression if there are data skew without shuffle. We need more designs to enable this improvement, like checking the ndv.

### Why are the changes needed?

Avoid perf regression

### Does this PR introduce _any_ user-facing change?

no

### CODE REVIEW

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q17.sf100/simplified.txt [hunk 1]
```diff
@@ -1,90 +1,97 @@
 TakeOrderedAndProject [i_item_id,i_item_desc,s_state,store_sales_quantitycount,store_sales_quantityave,store_sales_quantitystdev,store_sales_quantitycov,as_store_returns_quantitycount,as_store_returns_quantityave,as_store_returns_quantitystdev,store_returns_quantitycov,catalog_sales_quantitycount,catalog_sales_quantityave,catalog_sales_quantitystdev,catalog_sales_quantitycov]
```
Comment (cloud-fan) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q17.sf100/simplified.txt:1]: I checked it locally. Now the plan golden files are exactly the same with before #32875

**Comment by cloud-fan** [comment]
cc @sunchao @c21

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 2]
```diff
@@ -451,30 +468,24 @@ case class HashShuffleSpec(
...
   }
 
-  override def canCreatePartitioning: Boolean = true
+  override def canCreatePartitioning: Boolean = {
+    // To avoid potential data skew, we don't allow `HashShuffleSpec` to create partitioning if
+    // the hash partition keys are not the full join keys (the cluster keys). Then the planner
+    // will add shuffles with the default partitioning of `ClusteredDistribution`, which uses all
+    // the join keys.
+    if (SQLConf.get.getConf(SQLConf.REQUIRE_ALL_JOIN_KEYS_AS_PARTITION_KEYS)) {
+      distribution.clustering.forall(x => partitioning.expressions.exists(_.semanticEquals(x)))
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:477]: Do we need to require `partitioning.expressions` to be exactly same with `distribution.clustering` as well? e.g. for followed cases:

```
partitioning.expressions: [a, b]
distribution.clustering: [b, a]
```

```
partitioning.expressions: [a, b, a]
distribution.clustering: [a, b]
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -396,6 +396,16 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+  val REQUIRE_ALL_JOIN_KEYS_AS_PARTITION_KEYS =
+    buildConf("spark.sql.join.requireAllJoinKeysAsPartitionKeys")
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:400]: would this config take effect for all physical operators having 2 children and requiring `ClusteredDistribution`? Example like `CoGroupExec`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:477]: Good point. To fully restore to the previous behavior, we should require an exact match, though I think the current change should cover the data skew issues.

I'll make the change to be conservative.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:400]: Yes, and they use `HashClusteredDistribution` before.

Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:400]: Then should we make the config name more general?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:400]: how about `spark.sql.requireAllClusterKeysAsPartitionKeysToCoParition`?

Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:400]: SGTM!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 2 above)
Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:477]: Yes the case where partitioning expressions is a superset of distribution clustering should already be rejected by `HashPartitioning#satisfies`, but it maybe better to make it more explicit here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:400]: I'm slightly inclined to a shorter name like `spark.sql.join.requireAllJoinKeysAsPartitionKeys` but it's just a personal flavor. Also I'd suggest something like `spark.sql.enableStrictShuffleKeysCheck` but it's up to you.

**Review by sunchao — COMMENTED** [review]
Looks good to me. @cloud-fan did you actually observe the perf regression in real Spark jobs? just curious whether it'll be very common when this is config is disabled.

**Comment by cloud-fan** [comment]
@sunchao I haven't tried it on real workloads yet, but it's pretty obvious that we can construct a query with certain input data to expose this regression.

**Review by c21 — APPROVED** [review]
LGTM

**Review by sunchao — COMMENTED** [review]
@cloud-fan could you fix the test failure: looks like `org.apache.spark.sql.catalyst.ShuffleSpecSuite` is failing. LGTM after the test failures are addressed. I also compared the golden files with this PR and #32875 combined and there're no changes which is expected.

**Comment by c21** [comment]
Just two cents from our production experience. We enabled the same feature (avoid shuffle if bucket keys are subset of join keys) by default in our production for several years, and didn't see much data skew issues. Our workload is not representative in industry, but just to provide some observation in one large scale environment.

**Comment by sunchao** [comment]
Thanks @c21 ! this is good data point. We're also planning to evaluate this feature in production jobs.

**Comment by cloud-fan** [comment]
I think we can still roll out this optimization later, with some heuristics to avoid bad cases. We just need more time to evaluate and do experiments.

**Comment by sunchao** [comment]
Hmm `org.apache.spark.sql.sources.BucketedReadWithHiveSupportSuite` also failed.

**Comment by sunchao** [comment]
Oops we missed another one: "SPARK-27485: EnsureRequirements.reorder should handle duplicate expressions" in `PlannerSuite`

**Review by sigmod — APPROVED** [review]

**Pull request closed without merging**